libafl/src/events/mod.rs

```diff
@@ -588,80 +607,129 @@ pub trait HasEventManagerId {
     fn mgr_id(&self) -> EventManagerId;
 }
 
-/// [`EventManager`] is the main communications hub.
-/// For the "normal" multi-processed mode, you may want to look into [`LlmpRestartingEventManager`]
-pub trait EventManager<E, Z>:
-    EventFirer + EventProcessor<E, Z> + EventRestarter + HasEventManagerId + ProgressReporter
-where
-    Self::State: HasMetadata + HasExecutions + HasLastReportTime,
-{
-}
-
-/// The handler function for custom buffers exchanged via [`EventManager`]
+/// The handler function for custom buffers exchanged via [`EventManager`]. This signature matches
+/// [`CustomBufHandler::handle`].
+///
+/// In previous versions, you could dynamically add such fn-based handlers with the
+/// `add_custom_buf_handler` function. If you wish to continue to do so, use [`Vec::new`] as your
+/// [`CustomBufHandlerTuple`] in the associated `handlers` parameter for your event manager. You may
+/// then use [`SupportsDynamicHandlers::add_custom_buf_handler`].
+#[deprecated(
+    note = "Using function pointers may limit your ability to use some stages due to type conflicts regarding the state"
```

Comment:
Why deprecate?

Reply:
We should only deprecate stuff we want to drop, do we want to drop this? What's an example for such a type conflict?

Reply:
Because it is both potentially restrictive and bad practice. Deprecation is the only means by which we can communicate that to both new and old codebases at compile time.

Reply:
Why is it bad? If ppl want to use it and we support it we should not deprecated

Reply:
Or we don't support it anymore, then we should put a version into the deprecation note and remove it later

Reply:
i'll remove it

Reply:
Please don't!

Reply:
As noted in the deprecation: this restricts you to exactly one state, which might break things in very unexpected ways if someone uses e.g. sub-stages, changes stages mid-run, etc.

Reply:
There is a better way to do this, which is documented, such that you can avoid this problem.

Reply:
but only this?
can i delete custom_buf handlers from other structs?

Reply:
>  can i delete custom_buf handlers from other structs?

What exactly do you want to delete? If there's anything still explicitly using `Vec<Box<dyn ...>>>` then it should be changed to `HT`, like elsewhere.

Reply:
Isn't good documentation enough? The user will notice that it doesn't compile when it doesn't compile and then they can read the docs..

Reply:
```
    /// Create a event converter from a raw llmp client
    pub fn build_from_client<IC, ICB, SP>(
        self,
        llmp: LlmpClient<SP>,
        converter: Option<IC>,
        converter_back: Option<ICB>,
    ) -> Result<LlmpEventConverter<IC, ICB, SP, <??>>, Error>
    where 
        SP: ShMemProvider
    {
        Ok(LlmpEventConverter {
            throttle: self.throttle,
            last_sent: Duration::from_secs(0),
            llmp,
            #[cfg(feature = "llmp_compression")]
            compressor: GzipCompressor::with_threshold(COMPRESS_THRESHOLD),
            converter,
            converter_back,
            custom_buf_handlers: vec![],
        })
    }
```
how do i type this one?

Reply:
`Vec<CustomBufHandlerFn<S>>` and the generic should be something like `S: HasMetadata + HasNamedMetadata`?

Reply:
Or without metadata but at this point why even use the state :D

Reply:
no doesn't work @addisoncrump 
can you take a look
https://github.com/AFLplusplus/LibAFL/tree/no-uses-bbb

Reply:
i really want to delete this custom buf, it just offers the similar functionality as event hooks. so no reason to keep it

Reply:
> i really want to delete this custom buf, it just offers the similar functionality as event hooks. so no reason to keep it

If there is truly no functionality you cannot accomplish with event hooks instead, then I agree that all custom buf handlers should be removed. @domenukk for reference, you are the person who has the most use of this functionality iirc

Reply:
I would say "no", but historically @domenukk has disagreed. Apparently some people do some strange dynamic behaviour.

Reply:
Apart from being able to add and remove hooks on the fly, another benefit is that that dynamic buffers can be closures, static event hooks can not(I think?)
Capturing values takes away some headaches.
But yeah if you all want it gone we can remove it and worst case bring it back in the future...